### ISSUE TITLE
Automatic language classification for Untitled files

### ISSUE THREAD

**Issue by isidorn** [microsoft/vscode#118455]
We could do automatic language classification for Untitled files so users do not have to explicitly choose what language mode to use.

There are already modules for this in Python https://github.com/yoeo/guesslang
However it would be cool to have something in Javascript. One idea is to use TensurFlowJS and try to reuse the guesslang model.

Nice demo by Tyler:

https://user-images.githubusercontent.com/1926584/111809976-d18ca880-88d5-11eb-8e74-6c278aa2fd38.mov

**Comment by bpasero** [comment]
Some related issues: https://github.com/microsoft/vscode/labels/languages-guessing

**Comment by bpasero** [comment]
I would actually also use this for the CLI support for reading from stdin. Currently we create a tmp txt file that contains the contents that are piped into VS Code, but it would be awesome if we could change the language mode based on contents 👍

**Comment by isidorn** [comment]
@bpasero cool, glad there are more use cases.
Did not know we have this label :)

**Comment by isidorn** [comment]
@TylerLeonhardt did a great experiment, we plan to continue on this next milestone thus assigning to April
This PR has a checklist and discussion points https://github.com/microsoft/vscode/pull/119325

**Comment by TylerLeonhardt** [comment]
@bpasero I think for your request, if we fix this: https://github.com/microsoft/vscode/issues/41614

Then what your requesting could come "for free" if we add the logic to run the model on untitled files as I did in my prototype.

**Comment by bpasero** [comment]
Yeah unfortunately opening an untitled file from piping is currently not an option because we have no good way of talking from the one process directly into an untitled file. But you are probably right that this would be the right solution for that issue.

However, do we consider to support language detection also for plain text files? I wonder if the detection should also run when you open a txt file without having a language mode set. If we think that makes sense, it would solve the case for piping too.

**Comment by TylerLeonhardt** [comment]
In my eyes a .txt file is a plaintext file so I don't really think that scenario needs to be addressed but if we were to run the detect, we'd need to:

* change the file extension as well
* maybe run it less often (we said once every 100ish chars for untitled) since it's less likely that the user wants the detect logic than in the untitled case (then again running the model might not be intensive enough to need to do this)

**Comment by bpasero** [comment]
I am not sure the file extension needs to change, we can change the language today in any file without the need to change the file extension.

**Comment by TylerLeonhardt** [comment]
> I am not sure the file extension needs to change, we can change the language today in any file without the need to change the file extension.

I'm not sure all language runtimes can handle non-expected file extensions (PowerShell can't, for example) so if you try to F5 debug a file, and that file is a `.txt` file even though the language mode is set to `Python/PowerShell/TypeScript`, the debugger might get confused.

I'd want the language mode to change and then F5 debugging/running in their terminal (`node foo.js`) to all work flawlessly and leaving it as a txt might prevent that/cause confusion.

**Comment by bpasero** [comment]
Yeah I think that is fine, I actually think https://github.com/microsoft/vscode/issues/41614 is doable with some file watching tricks so I guess it is another good reason to look into it eventually.

**Comment by isidorn** [comment]
Assigning also @TylerLeonhardt since I know he is interested and half of June and most of July I will be on vacation.

**Comment by TylerLeonhardt** [comment]
A couple of pre-reqs:

- [ ] Figure out how to convert the native model to the tfjs model https://github.com/tensorflow/tfjs/issues/4838#issuecomment-859258920
- [ ] Figure out how to get tfjs to load the model over something other than http https://github.com/tensorflow/tfjs/issues/4838#issuecomment-859275362

Also if you'd like to play with the experience today (not fine tuned, but works...) the code is published as an extension here:
https://marketplace.visualstudio.com/items?itemName=TylerLeonhardt.auto-language

**Comment by TylerLeonhardt** [comment]
Moving this to July as I think there are few things we are blocked on above that I don't think we'll figure out in the next week or so.

**Comment by 4086606** [comment]
Is it possible to increase the accuracy using language heuristics data from [github/linguist](https://github.com/github/linguist)?

**Comment by isidorn** [comment]
@4086606 Good idea. However that library is a ruby library so we would have a dependency on ruby, which is a no-go for us, since we want this working in the browser. Unless there is a js alternative?

**Comment by 4086606** [comment]
Should be able to compile the entire gem to WASM, but I do wonder about size

**Comment by isidorn** [comment]
If you succeed do let us know how it went and the size. Though I see this as step 2, only in the case that the initial model classification does not prove super accurate.

**Comment by 4086606** [comment]
It seems like we will need to take the YAML files from that library and reimplement the heuristics logic. You're right that it should be a step 2 seeing as it's rather blind and not as accurate as it could be..

**Comment by 4086606** [comment]
@isidorn didn't work out - [github/linguist](https://github.com/github/linguist) can only _disambiguate_ file extensions. Guesslang should be able to use the linguist samples however

**Comment by TylerLeonhardt** [comment]
I was able to address the pre-reqs and published a new version of my extension that doesn't make a single network request:
https://marketplace.visualstudio.com/items?itemName=TylerLeonhardt.auto-language

Next order of business is to understand how to bring this into Core.

**Comment by TylerLeonhardt** [comment]
Closing in favor of https://github.com/microsoft/vscode/issues/129004
